(GH-2615) Add BoltSpec documentation (#2649)
puppetlabs/bolt: documentation/boltspec_reference.md

# BoltSpec reference

The `BoltSpec` helper library includes several functions to help write unit
tests for plans. For more information about writing unit tests for plans, and
how to use these functions, see [Testing plans](testing_plans.md).

## Mocks

Mocks serve two purposes when you are writing tests for your plans: they allow

> assertions about executor functions invoked during a plan run and also allow you

Reviewer: ```suggestion
assertions that executor functions are invoked during a plan run and also allow you
```

> to set their return values. For example, you might use a mock to make an

Reviewer: ```suggestion
to mock their return values. For example, you might use a mock to make an
```

Reviewer: idk if this is just more confusing, like using a word in it's own definition :P

Author: I think using mock in two contexts here could be confusing, so I'll leave it as 'set'.

assertion that the `run_task` function is invoked exactly one time during a

> test.

Reviewer: ```suggestion
test, and/or that it returns 'foo' on stdout.
```

> ### `expect_command`

Reviewer: This page overall seems pretty verbose, and like it includes a lot of duplication of expectations. What do you think about having a separate `Expectations` page that lists all the expectations, and then each function can just list the expectations it accepts and link to the Expectations reference page?

Author: I don't think verbosity/length should be a concern for reference pages since they're not meant to read straight through and have a ToC for easy navigation. Some of the modifiers also have different behavior (`be_called_times`) or have different parameters (`always_return`, `return_for_targets`, `return`) depending on the mock/stub used. We _could_ enumerate the exceptions for each mock/stub under the modifier, but I think that would make the page more confusing than anything.

Reviewer 2: Yeah, I agree. I think the way we expect users to interact with these reference pages is by using the TOC or using CMD+F, and it's reasonable for reference pages to be fairly repetitive for the sake of completeness. I'm also a little wary of adding more reference pages than we have to cos the references section is starting to get kinda chonky.

  fails if the file is downloaded from a different list of targets.

  ```ruby
  allow_download('/var/log/kern.log').with_targets(['target1', 'target2', 'target3'])
  ```

- `with_params(parameters)`

  The [options](plan_functions.md#download-file) that can be passed to the
  `download_file` function. The test fails if the file is downloaded with a
  different set of options.

  ```ruby
  allow_download('/var/log/kern.log').with_params({ '_run_as' => 'root' })
  ```

- `with_destination(destination)`

  The destination path that the file is downloaded to. The test fails if the
  file is downloaded to a different location.

  ```ruby
  allow_download('/var/log/kern.log').with_destination('kernel')
  ```

- `return(&block) { |targets, source, destination, params| ... }`

  Invokes a block to construct the `ResultSet` returned by the `download_file`
  function.

  ```ruby
  allow_download('/var/log/kern.log').return do |targets, source, destination, params|
    results = targets.map do |target|
      Bolt::Result.new(target, value: { 'path' => File.join(destination, source) })
    end

    Bolt::ResultSet.new(results)
  end
  ```

- `error_with(error)`

>   Sets the error hash for each target's `Result` when the file is downloadedn.

Reviewer 2: ```suggestion
  Sets the error hash for each target's `Result` when the file is downloaded.
```

  ```

- `with_params(parameters)`

  The [options](plan_functions.md#upload-file) that can be passed to the
  `upload_file` function. The test fails if the file is uploaded with a
  different set of  options.

  ```ruby
  allow_upload('sshd_config').with_params({ '_run_as' => 'root' })
  ```

- `with_destination(destination)`

  The destination path that the file is uploaded to. The test fails if the
  file is uploaded to a different location.

  ```ruby
  allow_upload('sshd_config').with_destination('/etc/ssh/sshd_config')
  ```

- `return(&block) { |targets, source, destination, params| ... }`

  Invokes a block to construct the `ResultSet` returned by the `upload_file`
  function.

  ```ruby
  allow_upload('sshd_config').return do |targets, source, destination, params|
    results = targets.map do |target|
      Bolt::Result.new(target, value: { 'path' => File.join(destination, source) })
    end

    Bolt::ResultSet.new(results)
  end
  ```

- `error_with(error)`

  Sets the error hash for each target's `Result` when the file is uploaded.
  Returns a `Bolt::ResultSet` object.

  ```ruby
  allow_upload('sshd_config').error_with('msg' => 'Not authorized')

>   ```

Reviewer 2: Similar to the testing doc, I think it's nice to bookend these with a related info section:

```
📖 Related information

- For more information on using `BoltSpec`, see [Testing plans](testing_plans.md).
```

> # BoltSpec reference

Reviewer: Is there a reason not to include `execute_no_plan` and `execute_any_plan` here? [This PR](https://github.com/puppetlabs/bolt/commit/6fdbaf25f3d94d54cfff0e6605c6154e0365618b#diff-f84e1d7c038c6793ef936447c02fa9c43dbf03afe70213d0b3f7eb75dc71509bR176) includes an example and documentation.

Author: I did have that originally, but removed it cause it was just a duplicate of what was in `testing_plans.md`. I can add it back.

Author: Here it is in `testing_plans.md` btw: https://github.com/puppetlabs/bolt/pull/2649/files#diff-2c26325655047f2d520c02a0f4cb9101097e114b65d6db4ed0c551940b502e42R444

Reviewer 2: Yeah, I think it's best to have them in the reference as well for completeness. And then it probably wouldn't hurt to link to testing_plans from each
